kartothek/serialization/_generic.py

```diff
@@ -203,6 +203,49 @@ def filter_df_from_predicates(df, predicates, strict_date_types=False):
     return df[indexer]
 
 
+def _ensure_type_stability(array_like, value, strict_date_types):
+    # datetime is a subclass of date which is why we need this double check
+    date_type_to_cast = datetime.datetime if strict_date_types else datetime.date
+    if isinstance(value, date_type_to_cast):
+        value = pd.Timestamp(value).to_datetime64()
+    elif is_list_like(value) and len(value) and isinstance(value[0], date_type_to_cast):
+        value = [pd.Timestamp(val).to_datetime64() for val in value]
+
+    value_dtype = pd.Series(value).dtype
+
+    if pd.api.types.is_categorical(array_like):
+        array_value_type = array_like.cat.categories.dtype
+        if array_like.cat.categories.is_monotonic:
+            array_like = array_like.cat.as_ordered()
+        else:
+            array_like = array_like.cat.reorder_categories(
+                array_like.cat.categories.sort_values(), ordered=True
+            )
+    else:
+        array_value_type = array_like.dtype
+
+    # NULL types might not be preserved well, so try to cast floats (pandas default type) to the value type
+    # Determine the type using the `kind` interface since this is common for a numpy array, pandas series and pandas extension arrays
+    if array_like.dtype.kind == "f" and np.isnan(array_like).all():
+        if array_like.dtype.kind != value_dtype.kind:
+            array_like = array_like.astype(value_dtype)
+            array_value_type = array_like.dtype
+
+    type_comp = {value_dtype.kind, array_value_type.kind}
+    if (
+        len(type_comp) > 1
+        # UINT and INT are accepted
+        and not type_comp == {"u", "i"}
```

Comment:
This section is a bit weird. First of all, previously we allowed comparison between int and uint. with strict type checking this wouldn't be allowed which is why I included this section. 
The following two cases are dealing with the different kinds of string types there are in numpy (O | object, S | (byte-)string, U | Unicode).

The way it is implemented atm is essentially how we handled it before and I believe between these types we do have valid comparison operators which is why I would understand to keep these exceptions. I don't have a strong opinion about how strict we should be here and am open for discussion :)

Reply:
I don't have too much authority about this, but opening the door to comparing arrays with different types of only a _selection_ of type combinations (and not others) seems ambiguous.
For example, if I would be a user, and compare arrays of type object with string, successfully; I might also expect to be able to compare, for example, integers with booleans, or integers with floats.
My suggestion would thus be to either:
 - A) Have a clear definition of what types should be able to be compared to each other (perhaps there is some appropriate numpy/pandas function which does something similar and we could use?)
- B) Strictly disallow comparisons of different types.

Reply:
@fjetter what's actually the reason we're allowing non-strict comparisons?

Reply:
There is none. I essentially fixed the test cases and figured we should have a discussion about this.
At least for these types there is a non ambiguous definition of equality. 

FYI I believe the string cases are quite artificial. I only. Caught it with hypothesis and I'm not sure if we can construct these with parquet tbh

Reply:
Looking at the algorithm again, I think we should keep it as it is proposed here. The main issue (as far as I understood) is that Python doesn't have proper scalar types (e.g. unsigned ints) and therefore we must be a little bit forgiving when comparing a Python scalar with a NumPy array. So I would say: let's use this PR and add a tracking issue to remove the `strict_date_types` param and hard-default to `strict_date_types=True`, because I think there is no good reason for `strict_date_types=False` (aka being more forgiving than really required).

Reply:
Yes it boils down to python not having a proper scalar typing system and I'd like to not have to do something like `("uint_col", "==", np.uint(1))`, especially considering that _our type model_ is arrow and not numpy

> remove the strict_date_types param 

I believe we'll need the parameter since we support `date_as_object` which affects this setting if we want to keep the definition of the predicates independent from this toggle. 
I agree that we should switch the default behavior

Reply:
> I believe we'll need the parameter since we support `date_as_object` which affects this setting
A right, I forgot. Out of interest: did I understand it correctly that this setting isn't required anymore with pyarrow 0.14?